File: Backtrack/78.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
    	List<List<Integer>> result = new ArrayList<>();
    	Arrays.sort(nums);
    	doBacktrack(result, new ArrayList<>(), nums, 0);
    	
    	return result;
    }

	private void doBacktrack(List<List<Integer>> result, List<Integer> temp, int[] nums, int start) {
		
		result.add(new ArrayList<>(temp));
		
		for (int i = start; i < nums.length; i++) {
			temp.add(nums[i]);
			doBacktrack(result, temp, nums, i + 1);
			temp.remove(temp.size() - 1);
		}
	}
}

File: 2024_may/78.cs
public class Solution
{
    public IList<IList<int>> Subsets(int[] nums)
    {
        IList<IList<int>> result = new List<IList<int>>();
        DoBacktrack(result, [], nums, 0);

        return result;
    }

    private void DoBacktrack(IList<IList<int>> result, List<int> temp, int[] nums, int start)
    {
        result.Add(new List<int>(temp));

        for (int i = start; i < nums.Length; i++)
        {
            temp.Add(nums[i]);
            DoBacktrack(result, temp, nums, i + 1);
            temp.RemoveAt(temp.Count - 1);
        }
    }
}